### PR TITLE
Extend detector `FindReturnRef` to detect exposition of references and buffers in nested and inner classes

### PR DESCRIPTION
**Pull request by baloghadamsoftware**
This patch fixes issue #2042

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [X] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java [hunk 1]
```diff
@@ -115,153 +123,206 @@ public void visit(Method obj) {
...
-        if (check) {
+        if (check || record) {
             super.visit(obj);
         }
     }
 
     @Override
     public void sawOpcode(int seen) {
 
-        if (!check) {
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java:126]: Why was this guard clause flipped? It's altered the indentation of the entire (very long) method, which makes it harder to identify the real changes.

Comment (JuditKnoll) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java:126]: This guard clause got flipped back, so it's not a problem anymore for the github diff to handle. IMO the code itself was a  bit more logical, but this way it's easy to see the changes.

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Review by JuditKnoll — APPROVED** [review]
This PR seems to solve the referenced issue.

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java [hunk 2]
```diff
@@ -80,13 +84,16 @@ public class FindReturnRef extends OpcodeStackDetector {
...
+    private final Map<XMethod, XField> accessMethodFields = new HashMap<>();
+
     private final BugAccumulator bugAccumulator;
 
     private static final Matcher BUFFER_CLASS_MATCHER = Pattern.compile("Ljava/nio/[A-Za-z]+Buffer;").matcher("");
     private static final Matcher DUPLICATE_METHODS_SIGNATURE_MATCHER =
             Pattern.compile("\\(\\)Ljava/nio/[A-Za-z]+Buffer;").matcher("");
     private static final Matcher WRAP_METHOD_SIGNATURE_MATCHER =
             Pattern.compile("\\(\\[.\\)Ljava/nio/[A-Za-z]+Buffer;").matcher("");
+    private static final Matcher ACCESS_METHOD_MATCHER = Pattern.compile("access\\$\\d{3}").matcher("");
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java:96]: `Matcher` instances are not thread-safe. This should store the `Pattern` and then create a `Matcher` for each string, not reset the same Matcher.

**Review by ThrawnCA — CHANGES_REQUESTED** [review]

**Comment by JuditKnoll** [comment]
Taken over by https://github.com/spotbugs/spotbugs/pull/3174.

**Pull request closed without merging**